pkg/river/vm/op_binary.go

```diff
@@ -0,0 +1,324 @@
+		// different Go type as needed (so that 3 == 3.0 is true).
+		lhsNum, rhsNum := lhs.Number(), rhs.Number()
+		switch fitNumberKinds(lhsNum.Kind(), rhsNum.Kind()) {
+		case value.NumberKindUint:
+			return lhsNum.Uint() == rhsNum.Uint()
+		case value.NumberKindInt:
+			return lhsNum.Int() == rhsNum.Int()
+		case value.NumberKindFloat:
+			return lhsNum.Float() == rhsNum.Float()
+		}
+
+	case value.TypeString:
+		return lhs.Text() == rhs.Text()
+
+	case value.TypeBool:
+		return lhs.Bool() == rhs.Bool()
+
+	case value.TypeArray:
+		// Two arrays are equal if they have equal elements.
+		if lhs.Len() != rhs.Len() {
+			return false
+		}
+		for i := 0; i < lhs.Len(); i++ {
+			if !valuesEqual(lhs.Index(i), rhs.Index(i)) {
+				return false
+			}
+		}
+		return true
+
+	case value.TypeObject:
+		// Two objects are equal if they have equal elements.
+		if lhs.Len() != rhs.Len() {
+			return false
+		}
+		for _, key := range lhs.Keys() {
+			lhsElement, _ := lhs.Key(key)
+			rhsElement, inRHS := rhs.Key(key)
+			if !inRHS {
+				return false
+			}
+			if !valuesEqual(lhsElement, rhsElement) {
+				return false
+			}
+		}
+		return true
+
+	case value.TypeFunction:
+		// Two functions are never equal. We can't compare functions in Go, so
+		// there's no way to compare them in River right now.
+		return false
+
+	case value.TypeCapsule:
+		// Two capsules are only equal if the underlying values are deeply equal.
+		return reflect.DeepEqual(lhs.Interface(), rhs.Interface())
+	}
+
+	panic("river/vm: unreachable")
+}
+
+// binopAllowedTypes maps what type of values are permitted for a specific
+// binary operation.
+//
+// token.EQ and token.NEQ are not included as they're handled separately from
+// other binary ops.
+var binopAllowedTypes = map[token.Token][]value.Type{
+	token.OR:  {value.TypeBool},
+	token.AND: {value.TypeBool},
+
+	token.ADD: {value.TypeNumber, value.TypeString},
+	token.SUB: {value.TypeNumber},
+	token.MUL: {value.TypeNumber},
+	token.DIV: {value.TypeNumber},
+	token.MOD: {value.TypeNumber},
+	token.POW: {value.TypeNumber},
+
+	token.LT:  {value.TypeNumber, value.TypeString},
+	token.GT:  {value.TypeNumber, value.TypeString},
+	token.LTE: {value.TypeNumber, value.TypeString},
+	token.GTE: {value.TypeNumber, value.TypeString},
+}
+
+func acceptableBinopType(val value.Value, op token.Token) bool {
+	allowed, ok := binopAllowedTypes[op]
+	if !ok {
+		panic("river/vm: unexpected binop type")
+	}
+
+	actualType := val.Type()
+	for _, allowType := range allowed {
+		if allowType == actualType {
+			return true
+		}
+	}
+	return false
+}
+
+func fitNumberKinds(a, b value.NumberKind) value.NumberKind {
+	aPrec, bPrec := numberKindPrec[a], numberKindPrec[b]
+	if aPrec > bPrec {
+		return a
+	}
+	return b
+}
+
+var numberKindPrec = map[value.NumberKind]int{
+	value.NumberKindUint:  0,
+	value.NumberKindInt:   1,
+	value.NumberKindFloat: 2,
+}
+
+func intPow[Number int64 | uint64](n, m Number) Number {
```

Comment:
Nice sneak in here 😅

Reply:
This seems like one of the cases where generics are useful :)